digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/service/lane/LaneService.scala

```diff
@@ -260,123 +250,103 @@ trait LaneOperations {
-                                        val isSideCodeAgainstWrong = lane.sideCode != mainLane21SideCode.value && lane.sideCode != SideCode.BothDirections.value
-
-                                        val isLaneToRemove = (isLaneCodeTowards && isSideCodeTowardsWrong ) || (isLaneCodeAgainst && isSideCodeAgainstWrong)
-
-                                        isValidLaneId && isLaneToRemove
-                                     }
-                                    .map(_.id)
-
-        val lanesToAdd = (lanes ++ toAdd).filterNot(lane => toRemove.contains(lane.id))
+        val updatedLanes = (lanes ++ toAdd).filterNot(lane => adjustedSideCode.exists(_.id == lane.id)) ++ adjustedSideCode
 
-        val newChangeSet = changeSet.copy(generatedPersistedLanes = changeSet.generatedPersistedLanes ++ lanesToAdd,
-                                    expiredLaneIds = changeSet.expiredLaneIds ++ toRemove,
-                                    adjustedSideCodes = changeSet.adjustedSideCodes ++ needAdjustSideCode )
-
-        (lanesToAdd, newChangeSet)
+        val lanesWithSideCodeAdjustment = adjustedSideCode.map(lane => SideCodeAdjustment(lane.id, SideCode(lane.sideCode)))
+        val newChangeSet = changeSet.copy(generatedPersistedLanes = updatedLanes,
+          adjustedSideCodes = changeSet.adjustedSideCodes ++ lanesWithSideCodeAdjustment)
 
+        (updatedLanes, newChangeSet)
 
       case TrafficDirection.TowardsDigitizing | TrafficDirection.AgainstDigitizing  =>
 
         val toAdd = if (!lanesToProcess.exists(lane => lane.laneCode == mainLaneDirectionOK))
-                      Seq(createPersistedLane(mainLaneDirectionOK, SideCode.BothDirections.value, baseLane.municipalityCode, baseProps))
-                    else
-                      Seq()
+          Seq(createPersistedLane(mainLaneDirectionOK, SideCode.BothDirections.value, baseLane.municipalityCode, baseProps))
+        else
+          Seq()
 
         val (laneCodeStart, laneCodeEnd) = getLaneNumbersRange(mainLaneDirectionOK)
-        val needAdjustSideCode = getLaneCodesForSideCodeAdjust(lanesToProcess, laneCodeStart, laneCodeEnd, SideCode.BothDirections)
 
-        val toRemove = lanesToProcess.filter { lane =>
-                                      val isValidLaneId = lane.id != 0L
-                                      val isLaneCodeToRemove = lane.laneCode >= mainLaneDirectionRemove._1 && lane.laneCode <= mainLaneDirectionRemove._2
-                                      val isSideCodeWrong = lane.sideCode != SideCode.BothDirections.value
+        //Adjust SideCode for lanes with correct laneCodes but wrong sideCodes
+        val lanesAdjustedNewSideCode = editLanesNeedSideCodeAdjust(lanesToProcess, laneCodeStart, laneCodeEnd, SideCode.BothDirections)
+        val lanesWithSideCodeAdjustment = lanesAdjustedNewSideCode.map(lane => SideCodeAdjustment(lane.id, SideCode.BothDirections))
```

Comment:
If Traffic Direction changes (towards to against or vice-versa) then this attempt to fix the Side Code will not do.
What seems to happen is the lane will be removed and new one is created.

Reply:
How so?
I agree with the extra process you mentioned since if there is a toAdd there is no need to adjust sidecodes because it will return empty.

Reply:
Maybe that parte can go inside the else part of the 
if (!lanesToProcess.exists(lane => lane.laneCode == mainLaneDirectionOK))
And instead  return 1 parameter only, maybe returning 2 parameters it is ok.

Other thing is  editLanesNeedSideCodeAdjust seems not be a good replacement of the previous function getLaneCodesForSideCodeAdjust.
Because in the getLaneCodesForSideCodeAdjust you will have the lanes as SideCodeAdjustment and with editLanesNeedSideCodeAdjust you are doing the same but here in this parte of the code.
Even for ' (lanes ++ toAdd).filterNot.....' you can use the result as SideCodeAdjustment.